fgspectra/models.py

```diff
@@ -1,89 +1,134 @@
 
-        missing_params = [p for p in params if params[p] is None]
-        if len(missing_params) > 0:
-            raise Exception(f'Missing parameters: {missing_params}.')
+    where :math:`k` and :math:`n` are component indices.
 
-        return params
+    Parameters
+    ----------
+    sed : callable
+        :math:`f(\nu)`. It returns an array with shape ``(comp, ..., freq)``.
+        It can be :class:`fgspectra.frequency.SED`.
+    cl_args : callable
+        :math:`C_\ell`. It returns an array with shape
+        ``(..., comp, comp, ell)``.
+        It can be :class:`fgspectra.power.PowerSpectrum`
 
-    def get_mix(self, freqs, l_max, **kwargs):
-        """Compute the mixing matrix up to some ell, with effective frequencies.
+    Note
+    ----
+    The two (optional) sets of extra dimensions ``...`` must be
+    broadcast-compatible.
+    """
+
+    def __init__(self, sed, cl):
+        self._sed = sed
+        self._cl = cl
+
+    def __call__(self, sed_args, cl_args):
+        """Compute the model at frequency and ell combinations.
 
         Parameters
         ----------
-        freqs : array of floats
-            Compute the mixing matrix containing cross-spectra at single
-            frequencies, i.e. delta function bandpowers. For example,
-            ACT-E has frequencies=[148.0, 220.0].
-
-        l_max : int
-            Maximum multipole to compute mixing matrix.
+        sed_args : list
+            Arguments for which the `sed` is evaluated.
+        cl_args : list
+            Arguments for which the `cl` is evaluated.
 
         Returns
         -------
-        3D array, first two dimensions correspond to frequency index
-        and third dimension is multipole.
+        cross : ndarray
+            Cross-spectrum. The shape is ``(..., freq, freq, ell)``.
         """
-        N_freq = len(freqs)
-        ells = np.arange(l_max+1)
-        mix = np.zeros((N_freq, N_freq, l_max+1))
-        for i in range(N_freq):
-            for j in range(N_freq):
-                params = kwargs.copy()
-                params['i'] = i
-                params['j'] = j
-                params['effective_frequencies'] = True
-                mix[i,j,:] = self.model(freqs[i], freqs[j], ells, **params)
+        f_nu = self._sed(*sed_args)
+        return np.einsum('k...i,n...j,...knl->...ijl',
+                         f_nu, f_nu, self._cl(*cl_args))
 
-        return mix
 
-    def get_mix_bandpowers(self, bandpower_windows, bandpower_frequencies):
-        raise NotImplementedError # gotta put in bandpower integration
+class PowerLaw(FactorizedCrossSpectrum):
+    """ Single Power law in both fequency and multipoles
+
+    See :class:`fgspectra.frequency.PowerLaw` for the frequency dependence and
+    :class:`fgspectra.power.PowerLaw` for the ell-dependence.
+    """
+
+    def __init__(self):
+        super().__init__(fgf.PowerLaw(), fgp.PowerLaw())
 
 
+'''
```

Comment:
This long list of subclasses is indeed super ugly! I am in favor of removing them from `models.py`. However, I'd really like fgspectra to be able to create/edit published models (i.e. Dunkley et al. 2013 or Planck XI 2015) with a few lines of code. For example, one use case is this [ACT multifrequency likelihood](https://github.com/xzackli/mfpy/blob/master/mfpy.py#L90), which has a call in its `__init__`, 

```python
        self.fg_components = [
            fgspectra.models.ThermalSZ(),
            fgspectra.models.KinematicSZ(),
            fgspectra.models.CIBP(),
            fgspectra.models.CIBC(),
            fgspectra.models.tSZxCIB(),
            fgspectra.models.RadioPointSources(),
            fgspectra.models.GalacticCirrus()
]
```
(this was the interface for fgspectra 0.1, which was hastily put together to show at the ACT meeting). Maybe I should take the dunkley13 models and move them into factory functions?

Reply:
I fully agree that the list should be in here somehow. I commented it out just because I didn't think about how yet.
The major change I'd introduce is putting correleted components in one single class that inherits from `CorrelatedFactorizedCrossSpectrum`. This would require to write some `CorrelatedComponentsFromFiles` class in power and some class that stitches together seds of different components

Reply:
Yes exactly a CorrelatedComponents class would be great.